### ISSUE TITLE
Add outline view

### ISSUE THREAD

**Issue by sladage** [Dart-Code/Dart-Code#107]
Having a side panel with the current document hierarchy in it would be nice.

**Comment by DanTup** [comment]
This is a good idea, but it would be best implemented in VS Code. Code exposes APIs for us to provide this sort of info so that it can be rendered consistently across languages. I'm not even sure it would be possible for us to do (I can't see any APIs for creating new panels).

There's a feature request for it in Code here:

https://github.com/Microsoft/vscode/issues/5605

If they were to add it, we would definitely implement the provider to enable this for Dart.

In the meantime, there is a sort-of outline view (as mentioned in that case) you can access by pressing `Ctrl+Shift+O`.

I've subscribed to that issue and if it gets added to Code, we'll reopen this and implement. In the meantime, I hope the document symbol list (`Ctrl+Shift+O` may suffice).

**Comment by DanTup** [comment]
Probably makes sense to keep external issues open until they're dealt with to avoid duplicates being opened (and as a reminder there's something we want to do that is waiting on something else).

**Comment by DanTup** [comment]
Probably this can be done with this new API?

It's possible this could be done using the new tree API now:

https://github.com/Microsoft/vscode/issues/12163#issuecomment-256813193
https://github.com/Microsoft/vscode-extension-samples/tree/master/tree-explorer-sample

**Comment by DanTup** [comment]
The API in Code is only available in Insiders and has no shipped yet.

**Comment by devoncarew** [comment]
This API is available now (and has been improved a bit in the current Insiders). You'd define some json, and then an associated TreeDataProvider:

```
		"views": {
			"explorer": [
				{
					"id": "dartFileOutline",
					"name": "File Outline",
					"when": "resourceLangId == 'dart'"
				}
			]
		},
```

**Comment by DanTup** [comment]
I've been watching this.. someone has created an extension that uses the existing APIs:

https://marketplace.visualstudio.com/items?itemName=patrys.vscode-code-outline

So for now you can install that extension (note: I have no idea who the author is or how well it works!). It seems there are some real limitations for now, so I was planning on waiting until it'd been refined a little more and also whether it looked like users would rather install an extension like that which adds it for all languages, or have each languages service provide it.

**Comment by rbenson** [comment]
Not a user of ext, or 100% sure how I got here, but @DanTup, have you seen https://github.com/Microsoft/vscode-extension-samples/tree/master/tree-view-sample? Not a production ready tool, but may help some?

**Comment by DanTup** [comment]
Yeah, I've seen that - it was used to build the extension linked above. I need to have a play with the API to see how well it works - last I read there were some limitations of the API but I haven't read the release notes for 1.14 so I don't know if some might've been addressed :-)

**Comment by DanTup** [comment]
This week's VS Code has added an Outline preview. It currently renders Dart's data flat (due to a mismatch in location information) but that's fixed as part of #986 for v2.14. Therefore I'm considering this (finally) done!
